samza-core/src/main/java/org/apache/samza/metrics/reporter/MetricsSnapshotReporter.java

```diff
@@ -0,0 +1,228 @@
+      // Ignore all exceptions - because subsequent executions of this scheduled task will be suppressed
+      // by the executor if the current task throws an unhandled exception.
+      LOG.warn("Error while reporting metrics. Will retry in " + reportingInterval + " seconds.", e);
+    }
+  }
+
+  public void innerRun() {
+    LOG.debug("Begin flushing metrics.");
+    for (RegistryWithSource registryWithSource : this.registries) {
+      String source = registryWithSource.getSource();
+      ReadableMetricsRegistry registry = registryWithSource.getRegistry();
+      LOG.debug("Flushing metrics for {}.", source);
+      Map<String, Map<String, Object>> metricsMsg = new HashMap<>();
+
+      // metrics
+      registry.getGroups().forEach(group -> {
+        Map<String, Object> groupMsg = new HashMap<>();
+        registry.getGroup(group).forEach((name, metric) -> {
+          if (!shouldIgnore(group, name)) {
+            metric.visit(new MetricsVisitor() {
+              @Override
+              public void counter(Counter counter) {
+                groupMsg.put(name, counter.getCount());
+              }
+
+              @Override
+              public <T> void gauge(Gauge<T> gauge) {
+                groupMsg.put(name, gauge.getValue());
+              }
+
+              @Override
+              public void timer(Timer timer) {
+                groupMsg.put(name, timer.getSnapshot().getAverage());
+              }
+            });
+          }
+        });
+
+        // dont emit empty groups
+        if (!groupMsg.isEmpty()) {
+          metricsMsg.put(group, groupMsg);
+        }
+      });
+
+      // publish to Kafka only if the metricsMsg carries any metrics
+      if (!metricsMsg.isEmpty()) {
+        MetricsHeader header =
+            new MetricsHeader(this.jobName, this.jobId, this.containerName, this.execEnvironmentContainerId, source,
+                this.version, this.samzaVersion, this.host, this.clock.currentTimeMillis(), this.resetTime);
+        Metrics metrics = new Metrics(metricsMsg);
+        LOG.debug("Flushing metrics for {} to {} with header and map: header={}, map={}.", source, out,
+            header.getAsMap(), metrics.getAsMap());
+        MetricsSnapshot metricsSnapshot = new MetricsSnapshot(header, metrics);
+        Object maybeSerialized = (this.serializer != null) ? this.serializer.toBytes(metricsSnapshot) : metricsSnapshot;
+        try {
+          this.producer.send(source, new OutgoingMessageEnvelope(this.out, this.host, null, maybeSerialized));
+          // Always flush, since we don't want metrics to get batched up.
+          this.producer.flush(source);
+        } catch (Exception e) {
+          LOG.error(String.format("Exception when flushing metrics for source %s", source), e);
+        }
+      }
+    }
+    LOG.debug("Finished flushing metrics.");
+  }
+
+  protected boolean shouldIgnore(String group, String metricName) {
+    boolean isBlacklisted = this.blacklist.isPresent();
```

Comment:
The check for `isBlacklisted` in the `if` does short-circuit the flow, so it is already returning reasonably early.
Since this class caches the blacklisted metrics, the flow here is slightly more complicated than it initially seems. I also initially felt there was a chance to simplify the logic, but then I just did the straightforward conversion because of the caching flow.